README.md
# Gleez Gravatar

[![Build Status](https://travis-ci.org/sergeyklay/gleez-gravatar.svg?branch=master)](https://travis-ci.org/sergeyklay/gleez-gravatar)
[![Dependency Status](https://www.versioneye.com/user/projects/5363c11efe0d07194f000001/badge.png)](https://www.versioneye.com/user/projects/5363c11efe0d07194f000001)


## What Is Gleez Gravatar?

[Gravatar's](http://gravatar.com/) are universal avatars available to all web sites and services. Users must register their email addresses with Gravatar before their avatars will be usable in your project.

The Gleez Gravatar component provides an easy way to retrieve a user's profile image from [Gravatar](http://gravatar.com/) based on a given email address. If the email address cannot be matched with a Gravatar account, an alternative will be returned based on the `Gravatar::$defaultImage` setting.

Users with gravatars can have a default image if you want to.


## Requirements

### Tested PHP versions

* 5.3.x
* 5.4.x
* 5.5.x
* 5.6.x
* hhvm

### Additional requirements

* PHPUnit 4.0.0 or higher (for testing)


## Installation

__Git way__

```bash
$ git clone https://github.com/sergeyklay/gleez-gravatar
```

__Composer way__

Download and install composer from [http://www.getcomposer.org](http://www.getcomposer.org):

```bash
$ curl -sS https://getcomposer.org/installer | php
```

Add the following to your project `composer.json` file:

```json
{
    "require": {
        "gleez/gravatar": "dev-master"
    }
}
```

When you're done just run


```bash
$ php composer.phar install
```

and the package is ready to be used. Another way, if you already have composer:

```bash
$ composer require --prefer-source gleez/gravatar:dev-master
```

__Direct zipball download__

https://github.com/sergeyklay/gleez-gravatar/archive/master.zip


## Getting Started

```php
// Some preparation with your class autoloader
// ...

// Get Gravatar instance
$gravatar = Gleez\Gravatar\Gravatar::getInstance();

// Setting default image, maximum size and maximum allowed Gravatar rating
$gravatar->setDefaultImage('retro')
         ->setSize(220)
         ->setRating('pg');

// Build the Gravatar URL based on the configuration and provided email address
$avatar = $gravatar->buildURL('john@doe.com');

```

## Configuration

### Gravatar size

Gravatar allows avatar images ranging from 0px to 2048px in size. With using 0px size, images from Gravatar will be returned as 80x80px.

Example:

```php
// Set gravatars size 64x64px
$gravatar->setSize(64);

// Set gravatars size 100x100px
$gravatar->setSize('100');
```

__WARNING__: If an invalid size (less than 0, greater than 2048) or a non-integer value is specified, this method will throw an exception of class `\InvalidArgumentException`.

### Default Image

Gravatar provides several pre-fabricated default images for use when the email address provided does not have a gravatar or when the gravatar specified exceeds your maximum allowed content rating. In addition, you can also set your own default image to be used by providing a valid URL to the image you wish to use.

There are a few conditions which must be met for default image URL:

* __MUST__ be publicly available (e.g. cannot be on an intranet, on a local development machine, behind HTTP Auth or some other firewall etc). Default images are passed through a security scan to avoid malicious content
* __MUST__ be accessible via HTTP or HTTPS on the standard ports, 80 and 443, respectively
* __MUST__ have a recognizable image extension (jpg, jpeg, gif, png)
* __MUST NOT__ include a query string (if it does, it will be ignored)

Possible values:

* `404` — do not load any image if none is associated with the email, instead return an HTTP 404 (File Not Found) response
* `mm` — (mystery-man) a simple, cartoon-style silhouetted outline of a person (does not vary by email)
* `identicon` — a geometric pattern based on an email
* `monsterid` — a generated 'monster' with different colors, faces, etc
* `wavatar` — generated faces with differing features and backgrounds
* `retro` — awesome generated, 8-bit arcade-style pixelated faces
* `blank` — a transparent PNG image
* boolean false — if using the default gravatar image
* Image URL

Example:

```php
// boolean false for the gravatar default
$gravatar->setDefaultImage(false);

// string specifying a recognized gravatar "default"
$gravatar->setDefaultImage('identicon');

// string with image URL
$gravatar->setDefaultImage("http://example.com/your-default-image.png");
```

__WARNING__: If an invalid default image is specified (both an invalid prefab default image and an invalid URL is provided), this method will throw an exception of class `\InvalidArgumentException`.

### Using secure connection

Should we use the secure (HTTPS) URL base? If your site is served over HTTPS, you'll likely want to serve gravatars over HTTPS as well to avoid "mixed content warnings". 

Example:

```php
// Enable secure connections:
$gravatar->enableSecureURL();

// Disable secure connections:
$gravatar->disableSecureURL();
```

To check to see if you are using "secure images" mode, call the method `Gravatar::useSecureURL()`, which will return a boolean value regarding whether or not secure images mode is enabled.

### Gravatar Rating

Gravatar provides four levels for rating avatars by, which are named similar to entertainment media ratings scales used in the United States.

Possible values:

* `g` — suitable for display on all websites with any audience type
* `pg` — may contain rude gestures, provocatively dressed individuals, the lesser swear words, or mild violence
* `r` — may contain such things as harsh profanity, intense violence, nudity, or hard drug use
* `x` — may contain hardcore sexual imagery or extremely disturbing violence

By default, the Gravater rating is `g`.

Example:

```php
$gravatar->setRating('PG');
```

__WARNING__: If an invalid maximum rating is specified, this method will throw an exception of class `\InvalidArgumentException`.

### Force Default

If for some reason you wanted to force the default image to always load, you can enable or disable it.

Example:

```php
// Enable
$gravatar->enableForceDefault();

// Disable
$gravatar->disableForceDefault();
```

To check to see if you are using "Force Default" mode, call the method `Gravatar::useForceDefault()`, which will return a boolean value.


## Contributing

So, you want to contribute to the Gleez Gravatar project? Well of course you do! I guess that's pretty clear. I mean, anyone could tell, right? 

Some things to watch out for:

* Please do PR to `next` branch
* Make sure that the code you write fits with the general style and coding standarts of the [Accepted PHP Standards](http://www.php-fig.org/psr/)
* Your code should be easy to understand, maintainable, and modularized
* Please _test_ your changes before submitting to make sure that not only they work, but have not broken other parts of the Gleez Gravatar
* If you do any API changes please follow [Semantic Versioning 2.0.0](http://semver.org/spec/v2.0.0.html)


## Tests

To run the tests at the command line, go to the `tests` directory and run `phpunit`:

```bash
$ cd tests
$ phpunit \
--coverage-text \
--coverage-clover ../build/logs/clover.xml \
--coverage-html   ../build/report
```

For additional information see [PHPUnit The Command-Line Test Runner](http://phpunit.de/manual/current/en/textui.html)


## License

Licensed under the [Gleez CMS License](http://gleezcms.org/license). © Gleez Technologies

CHANGELOG.md
* __1.0.0__

  * Initial Release


tests/bootstrap.php
<?php
/**
 * Gleez CMS (http://gleezcms.org)
 *
 * @link      https://github.com/gleez/cms Canonical source repository
 * @copyright Copyright (c) 2011-2014 Gleez Technologies
 * @license   http://gleezcms.org/license Gleez CMS License
 */

// turn on all errors
error_reporting(E_ALL);

// autoloader
require dirname(__DIR__) . '/autoload.php';


autoload.php
<?php
/**
 * Gleez CMS (http://gleezcms.org)
 *
 * @link      https://github.com/gleez/cms Canonical source repository
 * @copyright Copyright (c) 2011-2014 Gleez Technologies
 * @license   http://gleezcms.org/license Gleez CMS License
 */

spl_autoload_register(function ($class) {
    
    // what namespace prefix should be recognized?
    $prefix = 'Gleez\Gravatar\\';
    
    // does the requested class match the namespace prefix?
    $prefix_len = strlen($prefix);
    if (substr($class, 0, $prefix_len) !== $prefix) {
        return;
    }
    
    // strip the prefix off the class
    $class = substr($class, $prefix_len);
    
    // a partial filename
    $part = str_replace('\\', DIRECTORY_SEPARATOR, $class) . '.php';
    
    // directories where we can find classes
    $dirs = array(
        __DIR__ . DIRECTORY_SEPARATOR . 'src' . DIRECTORY_SEPARATOR . 'Gleez' . DIRECTORY_SEPARATOR . 'Gravatar',
        __DIR__ . DIRECTORY_SEPARATOR . 'tests' . DIRECTORY_SEPARATOR . 'src' . DIRECTORY_SEPARATOR . 'Gleez' . DIRECTORY_SEPARATOR . 'Gravatar',
    );
    
    // go through the directories to find classes
    foreach ($dirs as $dir) {
        $file = $dir . DIRECTORY_SEPARATOR . $part;
        if (is_readable($file)) {
            require $file;
            return;
        }
    }
});


src/Gleez/Gravatar/Gravatar.php
<?php
/**
 * Gleez CMS (http://gleezcms.org)
 *
 * @link      https://github.com/gleez/cms Canonical source repository
 * @copyright Copyright (c) 2011-2014 Gleez Technologies
 * @license   http://gleezcms.org/license Gleez CMS License
 */

namespace Gleez\Gravatar;

use InvalidArgumentException;

/**
 * The Gleez Gravatar component provides an easy way to retrieve a user's profile image
 * from Gravatar site based on a given email address.
 *
 * If the email address cannot be matched with a Gravatar account, an alternative will
 * be returned based on the Gravatar::$defaultImage. setting. Users with gravatars can
 * have a default image if you want to.
 *
 * @package  Gleez\Gravatar
 * @author   Gleez Team
 * @version  1.0.0
 */
class Gravatar implements Gravatarable
{
    /**
     * The gravatar service URL
     * @type string
     */
    const HTTP_URL  = 'http://www.gravatar.com/avatar/';

    /**
     * The secure gravatar service URL
     * @type string
     */
    const HTTPS_URL = 'https://secure.gravatar.com/avatar/';

    /**
     * Minimum size of requested avatar
     * @type int
     */
    const MIN_AVATAR_SIZE = 0;

    /**
     * Maximum size of requested avatar
     * @type int
     */
    const MAX_AVATAR_SIZE = 2048;

    /**
     * @var Gleez\Gravatar\Gravatar
     */
    protected static $instance;

    /**
     * The default image.
     *
     * Possible values:
     * - String of the gravatar-recognized default image "type" to use
     * - URL
     * - false if using the default gravatar image
     *
     * @var mixed
     */
    private $defaultImage = false;

    /**
     * Gravatar defaults
     * @var array
     */
    private $validDafaults = array(
        '404'       => true,
        'mm'        => true,
        'identicon' => true,
        'monsterid' => true,
        'wavatar'   => true,
        'retro'     => true,
        'blank'     => true
    );

    /**
     * Gravatar rating
     * @var array
     */
    private $validRatings = array(
        'g'  => true,
        'pg' => true,
        'r'  => true,
        'x'  => true
    );

    /**
     * The size to use for avatars
     * @var int
     */
    private $size = 80;

    /**
     * The maximum rating to allow for the avatar
     * @var string
     */
    private $rating = 'g';

    /**
     * Should we use the secure (HTTPS) URL base?
     * @var bool
     */
    private $secureURL = false;

    /**
     * The default image shall be shown even if user that has an gravatar profile.
     * @var bool
     */
    private $forceDefault = false;

    /**
     * The Class Constructor
     *
     * Private constructor because this is a singleton class.
     */
    private function __construct()
    {
    }

    /**
     * Create and return and return Gleez\Gravatar\Gravatar instance
     *
     * @return Gleez\Gravatar\Gravatar
     */
    public static function getInstance()
    {
        if (!static::$instance instanceof Gravatarable) {
            static::$instance = new static();
        }

        return static::$instance;
    }


    /**
     * Set the default image to use for avatars
     *
     * Possible $image formats:
     * - a string specifying a recognized gravatar "default"
     * - a string containing a valid image URL
     * - boolean false for the gravatar default
     *
     * @param mixed $image The default image to use
     * @return Gleez\Gravatar\Gravatar
     *
     * @throws \InvalidArgumentException
     */
    public function setDefaultImage($image)
    {
        if (false === $image) {
            $this->defaultImage = $image;

            return $this;
        }

        $default = strtolower(trim($image));

        if (!isset($this->validDafaults[$default])) {
            if(!filter_var($image, FILTER_VALIDATE_URL, FILTER_FLAG_HOST_REQUIRED)) {
                throw new InvalidArgumentException(
                    'The default image specified is not a recognized gravatar "default" and is not a valid URL'
                );
            } else {
                $this->defaultImage = $image;
            }
        } else {
            $this->defaultImage = $default;
        }

        return $this;
    }

    /**
     * Get the current default image
     *
     * @return string  if one is set
     * @return bool    false if no default image set
     */
    public function getDefaultImage()
    {
        return $this->defaultImage;
    }

    /**
     * Set the avatar size to use
     *
     * By default, images from Gravatar.com will be returned as 80x80px
     *
     * @param int $size The avatar size to use
     * @return Gleez\Gravatar\Gravatar
     *
     * @throws \InvalidArgumentException
     */
    public function setSize($size)
    {
        $options = array(
            'options' => array(
                'min_range' => static::MIN_AVATAR_SIZE,
                'max_range' => static::MAX_AVATAR_SIZE
        ));

        if(false === filter_var($size, FILTER_VALIDATE_INT, $options)) {
            throw new InvalidArgumentException(
                sprintf(
                    "Can't set Gravatar size. Size must be an integer within %s and %s pixels",
                    static::MIN_AVATAR_SIZE, static::MAX_AVATAR_SIZE
                )
            );
        }

        $this->size = $size;

        return $this;
    }

    /**
     * Get the currently set avatar size
     *
     * @return int
     */
    public function getSize()
    {
        return $this->size;
    }

    /**
     * Set the maximum allowed rating for avatars
     *
     * @param string $rating The maximum rating to use for avatars ('g', 'pg', 'r', 'x')
     * @return Gleez\Gravatar\Gravatar
     *
     * @throws \InvalidArgumentException
     */
    public function setRating($rating)
    {
        $rating = strtolower(trim($rating));
        $allowed = implode(', ', array_keys($this->validRatings));

        if (!isset($this->validRatings[$rating])) {
            throw new InvalidArgumentException(
                sprintf("Invalid rating '%s' specified. Available for use only: %s.", $rating, $allowed)
            );
        }

        $this->rating = $rating;

        return $this;
    }

    /**
     * Get the current maximum allowed rating for avatars
     *
     * The string representing the current maximum allowed rating ('g', 'pg', 'r', 'x').
     *
     * @return int
     */
    public function getRating()
    {
        return $this->rating;
    }

    /**
     * Check if we are using the secure protocol for the image URLs
     *
     * @return bool
     */
    public function useSecureURL()
    {
        return $this->secureURL;
    }

    /**
     * Enable the use of the secure protocol for image URLs
     *
     * @return Gleez\Gravatar\Gravatar
     */
    public function enableSecureURL()
    {
        $this->secureURL = true;

        return $this;
    }

    /**
     * Disable the use of the secure protocol for image URLs
     *
     * @return Gleez\Gravatar\Gravatar
     */
    public function disableSecureURL()
    {
        $this->secureURL = false;

        return $this;
    }

    /**
     * Get the email hash to use
     *
     * @param string $email The email to get the hash for
     * @return string
     */
    public function getEmailHash($email)
    {
        return md5(strtolower(trim($email)));
    }

    /**
     * Forces Gravatar to display default image
     *
     * @return Gleez\Gravatar\Gravatar
     */
    public function enableForceDefault()
    {
        $this->forceDefault = true;

        return $this;
    }

    /**
     * Disable forces default image
     *
     * @return Gleez\Gravatar\Gravatar
     */
    public function disableForceDefault()
    {
        $this->forceDefault = false;

        return $this;
    }

    /**
     * Check if need to force the default image to always load
     *
     * @return bool
     */
    public function useForceDefault()
    {
        return $this->forceDefault;
    }

    /**
     * Build the Gravatar URL based on the configuration and provided email address
     *
     * @param string $email The email to get the gravatar for
     * @return string
     */
    public function buildURL($email)
    {
        $url = static::HTTP_URL;

        if ($this->useSecureURL()) {
            $url = static::HTTPS_URL;
        }

        $url .= $this->getEmailHash($email);

        $query = array(
            's' => $this->getSize(),
            'r' => $this->getRating()
        );

        if ($this->getDefaultImage()) {
            $query = array_merge($query, array('d' => $this->getDefaultImage()));
        }

        if ($this->useForceDefault()) {
            $query = array_merge($query, array('f' => 'y'));
        }

        $url .= '?'.http_build_query($query, '', '&');

        return $url;
    }
}
